Aplib.Core/Belief/SampledMemoryBelief.cs

```diff
@@ -0,0 +1,123 @@
+using System;
+using static Aplib.Core.Belief.UpdateMode;
+
+namespace Aplib.Core.Belief
+{
+    /// <summary>
+    /// The <see cref="SampledMemoryBelief{TReference, TObservation}"/> class represents the agent's belief of a single object,
+    /// but with additional "memory" of previous observations.
+    /// These observations are sampled at a fixed rate.
+    /// Some <i>object reference</i> is used to generate/update an <i>observation</i> 
+    /// (i.e., some piece of information on the game state as perceived by an agent).
+    /// This belief also stores a limited amount of previous observation samples in memory.
+    /// Optionally, the belief can always store the most recent observation, regardless of the sample rate.
+    /// </summary>
+    /// <remarks>
+    /// It supports implicit conversion to <typeparamref name="TObservation"/>.
+    /// </remarks>
+    /// <typeparam name="TReference">The type of the reference used to generate/update the observation.</typeparam>
+    /// <typeparam name="TObservation">The type of the observation the belief represents.</typeparam>
+    public class SampledMemoryBelief<TReference, TObservation> : MemoryBelief<TReference, TObservation>
+    {
+        /// <summary>
+        /// The sample interval of the memory (inverse of the sample rate).
+        /// One observation memory (i.e., snapshot) is stored every <see cref="_sampleInterval"/>-th cycle.
+        /// </summary>
+        private readonly int _sampleInterval;
+
+        /// <summary>
+        /// Specifies how this sampled memory belief should be updated.
+        /// </summary>
+        private readonly UpdateMode _updateMode;
+
+        private int _counter = 0;
+
+        /// <summary>
+        /// The number of cycles that have passed since the last memory sample was stored.
+        /// </summary>
+        private int Counter
```

Comment:
Your comment says it all, you could rename this to `CycleCount`

Reply:
I will call it ModuloCounter, as that seems to be the more commonly used name for this behavior.